Resolve longest match first

Given:

```
moduleAlias.addAliases({
    "react-dom": "preact-compat-enzyme",
    "react-dom/server": "preact-render-to-string"
});
```

An attempt to `require('react-dom/server')` yields `Error: Cannot find module 'preact-compat-enzyme/server'`.

I believe the simple solution would be to evaluate possible matches, and choose the longest.

--- a/index.js
+++ b/index.js
@@ -5,6 +5,7 @@ var nodePath = require('path')
 
 var modulePaths = []
 var moduleAliases = {}
+var moduleAliasNames = [];
 
 var oldNodeModulePaths = Module._nodeModulePaths
 Module._nodeModulePaths = function (from) {
@@ -21,12 +22,15 @@ Module._nodeModulePaths = function (from) {
 
 var oldResolveFilename = Module._resolveFilename
 Module._resolveFilename = function (request, parent, isMain) {
-  for (var alias in moduleAliases) {
+  for (var i = moduleAliasNames.length; i-- > 0;) {
+    var alias = moduleAliasNames[i]
     if (isPathMatchesAlias(request, alias)) {
       request = nodePath.join(
         moduleAliases[alias],
         request.substr(alias.length)
       )
+      // Only use the first match
+      break
     }
   }
 
@@ -35,12 +39,9 @@ Module._resolveFilename = function (request, parent, isMain) {
 
 function isPathMatchesAlias (path, alias) {
   // Matching /^alias(\/|$)/
-  if (path.indexOf(alias) === 0) {
-    if (path.length === alias.length) return true
-    if (path[alias.length] === '/') return true
-  }
-
-  return false
+  return  path.indexOf(alias) === 0
+      && (path.length === alias.length
+        || path[alias.length] === '/')
 }
 
 function addPathHelper (path, targetArray) {
@@ -86,6 +87,9 @@ function addAliases (aliases) {
 
 function addAlias (alias, target) {
   moduleAliases[alias] = target
+  // Cost of sorting is lower here than during resolution
+  moduleAliasNames = Object.keys(moduleAliases)
+  moduleAliasNames.sort()
 }
 
 /**

--- a/test/specs.js
+++ b/test/specs.js
@@ -109,4 +109,20 @@ describe('module-alias', function () {
       done()
     })
   })
+
+  it('should match longest alias first', function () {
+    moduleAlias.addAliases({
+      'react-dom': path.join(__dirname, 'src/bar/baz'),
+      'react-dom/server': path.join(__dirname, 'src/foo')
+    })
+
+    var bar, src
+    try {
+      bar = require('react-dom')
+      src = require('react-dom/server')
+    } catch (e) {}
+
+    expect(bar).to.equal('Hello from baz')
+    expect(src).to.equal('Hello from foo')
+  })
 })

